File: _dsv/join.py
import argparse
import threading
from queue import Queue
from ._column_slicer import _ColumnSlicer
from . import _utils
from ._shtab import shtab

class join(_ColumnSlicer):
    ''' join lines of two files on a common field '''
    parser = argparse.ArgumentParser()
    parser.add_argument('file', type=argparse.FileType('rb'), metavar='FILE', help='join stdin with FILE').complete = shtab.FILE
    parser.add_argument('fields', nargs='*', help='join on these fields from stdin and FILE')
    parser.add_argument('-1', dest='left_fields', action='append', help='join on these fields from stdin')
    parser.add_argument('-2', dest='right_fields', action='append', help='join on these fields from FILE')
    parser.add_argument('-e', dest='empty_value', type=_utils.utf8_type, default='', metavar='STRING', help='replace missing input fields with STRING')
    parser.add_argument('-r', '--regex', action='store_true', help='treat fields as regexes')
    parser.add_argument('--rename-1', type=_utils.utf8_type, help='rename header from stdin according to this %%-format string')
    parser.add_argument('--rename-2', type=_utils.utf8_type, help='rename header from FILE according to this %%-format string')
    group = parser.add_mutually_exclusive_group()
    group.add_argument('-a', dest='show_all', choices=('1', '2'), action='append', help='also print unpairable lines from the given file')
    group.add_argument('--join', choices=('inner', 'left', 'right', 'outer'), default='inner', help='type of join to perform')
    group.add_argument('--inner', action='store_const', dest='join', const='inner', help='do a inner join')
    group.add_argument('--left', action='store_const', dest='join', const='left', help='do a left join')
    group.add_argument('--right', action='store_const', dest='join', const='right', help='do a right join')
    group.add_argument('--outer', action='store_const', dest='join', const='outer', help='do a outer join')

    def __init__(self, opts):
        opts.fields.extend(opts.extras)
        opts.extras = ()

        if opts.fields and (opts.left_fields or opts.right_fields):
            self.parser.error('Cannot set common fields and use -1 or -2')

        right_opts = argparse.Namespace(**vars(opts))
        right_opts.fields = right_opts.fields.copy() or right_opts.right_fields or []
        opts.fields = opts.fields or opts.left_fields or []

        if opts.show_all:
            if '1' in opts.show_all and '2' in opts.show_all:
                opts.join = 'outer'
            elif '1' in opts.show_all:
                opts.join = 'left'
            elif '2' in opts.show_all:
                opts.join = 'right'

        super().__init__(opts)
        self.left = {}
        self.right = Queue()
        self.header_lock = threading.Semaphore(1)
        self.header_event = threading.Event()
        self.joined_header = None

        self.collector = _ColumnSlicer(right_opts)
        self.collector.on_header = self.on_collector_header
        self.collector.on_row = self.right.put_nowait
        self.collector.on_eof = lambda: 0 # nop

        self.thread = threading.Thread(target=self.run_collector, daemon=True)
        self.thread.start()

    def run_collector(self):
        try:
            list(self.collector.process_file(self.opts.file))
        finally:
            self.right.put_nowait(None)

    def on_header(self, header):
        self.header_map = self.make_header_map(self.header)
        return self.got_header()

    def on_collector_header(self, header):
        self.collector.header_map = self.make_header_map(self.collector.header)
        return self.got_header()

    def got_header(self):
        no_join_fields = not self.opts.fields and not self.collector.opts.fields

        # the semaphore only has 1, so the second time it is called it will return false and go through
        if not self.header_lock.acquire(blocking=False):
            if no_join_fields:
                self.opts.fields = self.collector.opts.fields = list(set(self.header) & set(self.collector.header))

            left = self.header
            right = self.collector.header

            if self.opts.rename_1:
                left = [self.opts.rename_1 % h for h in left]
            if self.opts.rename_2:
                right = [self.opts.rename_2 % h for h in right]

            self.joined_header = self.paste_row(left, right)
            try:
                if super().on_header(self.joined_header):
                    return True
            finally:
                self.header_event.set()

        if no_join_fields:
            # wait for the other thread to set the join fields
            self.header_event.wait()

    def paste_row(self, left, right):
        return self.slice(left) + self.slice(left, True) + self.collector.slice(right, True)

    def on_row(self, row):
        key = tuple(self.slice(row, False))
        self.left.setdefault(key, []).append(row)

    def join_left_with_right(self):
        first_left = self.header or list(self.left.values() or [[]])[0]
        key_len = len(self.slice(first_left))
        left_len = len(self.slice(first_left, True))
        right_len = self.collector.header and len(self.collector.slice(self.collector.header, True))

        matched = set()

        # left has finished, now read off the right
        while (right := self.right.get()) is not None:

            if right_len is None:
                right_len = len(self.collector.slice(right, True))

            key = tuple(self.collector.slice(right, False))
            matched.add(key)
            lefts = self.left.get(key, ())

            key = list(key)
            if len(key) < key_len:
                key += [b''] * (key_len - len(key))

            # inner joins
            right = self.collector.slice(right, True)
            for left in lefts:
                row = self.slice(left, True)
                if len(row) < left_len:
                    row += [b''] * (left_len - len(row))

                if super().on_row(key + row + right):
                    return True

            # right joins
            if not lefts and self.opts.join in ('right', 'outer'):
                padding = [self.opts.empty_value] * left_len
                if super().on_row(key + padding + right):
                    return True

        if right_len is None:
            right_len = 0

        # left joins
        if self.opts.join in ('left', 'outer'):
            for key in self.left.keys() - matched:
                lefts = self.left[key]

                key = list(key)
                if len(key) < key_len:
                    key += [b''] * (key_len - len(key_len))

                padding = [self.opts.empty_value] * right_len
                for left in lefts:
                    row = key + self.slice(left, True)
                    if len(row) < left_len:
                        row += [b''] * (left_len - len(row))

                    if super().on_row(row + padding):
                        return True

    def on_eof(self):
        self.join_left_with_right()
        super().on_eof()
        self.thread.join()

File: src/subcommands/join.rs
use crate::utils::Break;
use crate::utils::MaybeBreak;
use anyhow::{Result, Context};
use std::sync::mpsc::{self, Sender, Receiver};
use crate::base::*;
use bstr::{BString};
use std::collections::{HashSet, HashMap, hash_map::Entry};
use crate::column_slicer::ColumnSlicer;
use clap::{Parser};

#[derive(Parser, Clone)]
#[command(about = "join lines of two files on a common field")]
pub struct Opts {
    #[arg(value_name = "FILE", help = "join stdin with FILE")]
    file: String,
    #[arg(conflicts_with_all = ["left_fields", "right_fields"], help = "join on these fields from stdin and FILE")]
    fields: Vec<String>,
    #[arg(short = '1', help = "join on these fields from stdin")]
    left_fields: Vec<String>,
    #[arg(short = '2', help = "join on these fields from FILE")]
    right_fields: Vec<String>,
    #[arg(short = 'e', value_name = "STRING", help = "replace missing input fields with STRING")]
    empty_value: Option<String>,
    #[arg(long, help = "treat fields as regexes")]
    regex: bool,
    #[arg(long, help = "rename header from stdin according to this %%-format string")]
    rename_1: Option<String>,
    #[arg(long, help = "rename header from FILE according to this %%-format string")]
    rename_2: Option<String>,
    #[arg(short = 'a', overrides_with_all = ["join", "inner", "left", "right", "outer"], value_parser = ["1", "2"], help = "also print unpairable lines from the given file")]
    show_all: Vec<String>,
    #[arg(long, value_parser = ["inner", "left", "right", "outer"], help = "type of join to perform")]
    join: Option<String>,
    #[arg(long, help = "do a inner join")]
    inner: bool,
    #[arg(long, help = "do a left join")]
    left: bool,
    #[arg(long, help = "do a right join")]
    right: bool,
    #[arg(long, help = "do a outer join")]
    outer: bool,
}

#[derive(Copy, Clone)]
enum Join {
    Inner,
    Left,
    Right,
    Outer,
}

type Row = Vec<BString>;

pub struct Handler {
    inner: Child,
    err_receiver: Receiver<Result<()>>,
}

impl Handler {
    pub fn new(mut opts: Opts, base: &mut Base) -> Result<Self> {
        if !opts.fields.is_empty() {
            opts.left_fields = opts.fields.clone();
            opts.right_fields = std::mem::take(&mut opts.fields);
        }

        let join = match opts.join.as_deref() {
            Some("inner") => Join::Inner,
            Some("left") => Join::Left,
            Some("right") => Join::Right,
            Some("outer") => Join::Outer,
            _ if opts.inner => Join::Inner,
            _ if opts.left => Join::Left,
            _ if opts.right => Join::Right,
            _ if opts.outer => Join::Outer,
            _ => match (opts.show_all.iter().any(|x| x == "1"), opts.show_all.iter().any(|x| x == "2")) {
                (true, true) => Join::Outer,
                (true, false) => Join::Left,
                (false, true) => Join::Right,
                (false, false) => Join::Inner,
            },
        };

        let (sender, receiver) = mpsc::channel();
        let (err_sender, err_receiver) = mpsc::channel();
        let child = Child{ got_header: false, left: true, sender: Some(sender) };

        // start a thread to join everything
        {
            let mut base = base.clone();
            let opts = opts.clone();
            let err_sender = err_sender.clone();
            base.scope.spawn(move || {
                let result = (|| {
                    Joiner::default().do_joining(join, &opts, receiver, &mut base)?;
                    base.on_eof()?;
                    Ok(())
                })();
                err_sender.send(result).unwrap();
            });
        }

        // start a thread to read from rhs
        {
            let mut base = base.clone();
            let right_file = std::mem::take(&mut opts.file);
            let mut child = child.clone();
            child.left = false;
            base.scope.spawn(move || {
                let result = (|| {
                    let file = std::fs::File::open(&right_file).with_context(|| format!("failed to open {right_file}"))?;
                    let file = std::io::BufReader::new(file);
                    child.process_file(file, &mut base, Callbacks::ON_HEADER | Callbacks::ON_ROW)?;
                    Ok(())
                })();
                err_sender.send(result).unwrap();
            });
        }

        // and we will read from lhs ...

        Ok(Self {
            inner: child,
            err_receiver,
        })
    }
}

impl Processor for Handler {
    fn on_row(&mut self, base: &mut Base, row: Vec<BString>) -> Result<()> {
        self.inner.on_row(base, row)
    }

    fn on_header(&mut self, base: &mut Base, header: Vec<BString>) -> Result<()> {
        self.inner.on_header(base, header)
    }

    fn on_ofs(&mut self, _base: &mut Base, ofs: Ofs) -> MaybeBreak {
        Break::when(self.inner.sender.as_ref().unwrap().send((true, Message::Ofs(ofs))).is_err())
    }

    fn on_eof(self, _base: &mut Base) -> Result<bool> {
        drop(self.inner.sender);
        crate::utils::chain_errors(self.err_receiver)?;
        Ok(false)
    }
}

#[derive(Default)]
struct Joiner {
    key_len: usize,
    left_len: usize,
    right_len: usize,
}

type JoinStore = HashMap<Row, Vec<Row>>;

impl Joiner {
    fn do_joining(&mut self, join: Join, opts: &Opts, receiver: Receiver<(bool, Message)>, base: &mut Base) -> Result<()> {
        let fields = (&opts.left_fields, &opts.right_fields);
        let is_fields_set = !fields.0.is_empty() || !fields.1.is_empty();
        let mut stores = (HashMap::new(), HashMap::new());
        let mut headers = (None, None);
        let mut slicers = (ColumnSlicer::new(fields.0, opts.regex), ColumnSlicer::new(fields.1, opts.regex));
        let mut got_headers = false;
        let mut buffer = vec![];

        for (is_left, msg) in &receiver {
            match msg {
                Message::Separator => unreachable!(),
                Message::Raw(..) => unreachable!(),
                Message::Eof => (),
                Message::Stderr(_) => unreachable!(),
                Message::RawStderr(..) => unreachable!(),
                Message::Ofs(ofs) => if is_left {
                    base.on_ofs(ofs)?;
                },
                Message::Header(header) => {
                    if is_left {
                        headers.0 = Some(header);
                    } else {
                        headers.1 = Some(header);
                    }

                    if !got_headers && let Some(headers) = headers.0.as_ref().zip(headers.1.as_ref()) {
                        got_headers = true;

                        if ! is_fields_set {
                            // get common fields
                            let left: HashSet<_> = headers.0.iter().collect();
                            let right: HashSet<_> = headers.1.iter().collect();
                            let fields: Vec<_> = left.intersection(&right).copied().cloned().collect();
                            if fields.is_empty() {
                                // default join field is the first
                                slicers.0 = ColumnSlicer::new(&["1".into()], false);
                                slicers.1 = ColumnSlicer::new(&["1".into()], false);
                            } else {
                                slicers.0 = ColumnSlicer::from_names(fields.iter());
                                slicers.1 = ColumnSlicer::from_names(fields.iter());
                            }
                        }

                        // make header maps
                        slicers.0.make_header_map(headers.0);
                        slicers.1.make_header_map(headers.1);

                        // paste the headers together
                        let mut header = slicers.0.slice(headers.0, false, true);
                        let mut left = slicers.0.slice(headers.0, true, true);
                        let mut right = slicers.1.slice(headers.1, true, true);

                        if let Some(rename) = &opts.rename_1 {
                            for h in &mut left {
                                *h = crate::utils::percent_format(rename.as_bytes().into(), h.as_ref());
                            }
                        }
                        if let Some(rename) = &opts.rename_2 {
                            for h in &mut right {
                                *h = crate::utils::percent_format(rename.as_bytes().into(), h.as_ref());
                            }
                        }

                        header.append(&mut left);
                        header.append(&mut right);
                        base.on_header(header)?;

                        // clear out the buffered rows
                        // their side must be the other side
                        for row in buffer.drain(..) {
                            if self.on_row(!is_left, row, &mut stores, &slicers, base)? {
                                return Ok(())
                            }
                        }
                    }
                },
                Message::Row(row) => {
                    if !got_headers {
                        // stick it in the buffer for later
                        buffer.push(row);
                    } else if self.on_row(is_left, row, &mut stores, &slicers, base)? {
                        return Ok(())
                    }
                },
            }
        }

        if matches!(join, Join::Left | Join::Outer) {
            for (key, rows) in &stores.0 {
                if !stores.1.contains_key(key) {
                    for row in rows {
                        let row = self.make_row(key, Some(row), None, &slicers, opts.empty_value.as_ref());
                        base.on_row(row)?;
                    }
                }
            }
        }

        if matches!(join, Join::Right | Join::Outer) {
            for (key, rows) in &stores.1 {
                if !stores.0.contains_key(key) {
                    for row in rows {
                        let row = self.make_row(key, None, Some(row), &slicers, opts.empty_value.as_ref());
                        base.on_row(row)?;
                    }
                }
            }
        }
        Ok(())

    }

    fn make_row(
        &self,
        key: &Row,
        left: Option<&Row>,
        right: Option<&Row>,
        slicers: &(ColumnSlicer, ColumnSlicer),
        empty_value: Option<&String>,
    ) -> Row {
        let mut new_row = key.clone();
        new_row.resize(new_row.len().max(self.key_len), b"".into());

        let old_len = new_row.len();
        if let Some(left) = left {
            new_row.append(&mut slicers.0.slice(left, true, true));
        }
        let empty = empty_value.filter(|_| left.is_none()).map_or(b"" as _, |x| x.as_bytes());
        new_row.resize(new_row.len().max(old_len + self.left_len), empty.into());

        let old_len = new_row.len();
        if let Some(right) = right {
            new_row.append(&mut slicers.1.slice(right, true, true));
        } else if let Some(empty) = empty_value {
            new_row.resize(new_row.len().max(old_len + self.right_len), empty.as_bytes().into());
        }

        new_row
    }

    fn on_row(
        &mut self,
        is_left: bool,
        row: Row,
        stores: &mut (JoinStore, JoinStore),
        slicers: &(ColumnSlicer, ColumnSlicer),
        base: &mut Base,
    ) -> Result<bool> {

        if is_left && self.key_len == 0 {
            self.key_len = slicers.0.slice(&row, false, true).len();
        }
        if is_left && self.left_len == 0 {
            self.left_len = slicers.0.slice(&row, true, true).len();
        }
        if !is_left && self.right_len == 0 {
            self.right_len = slicers.1.slice(&row, true, true).len();
        }

        let left = (&slicers.0, &mut stores.0);
        let right = (&slicers.1, &mut stores.1);
        let (mut this, mut other) = (left, right);
        if !is_left {
            (this, other) = (other, this);
        }

        let key = this.0.slice(&row, false, true);
        // find any joined rows
        if let Some(other_rows) = other.1.get(&key) {
            for other_row in other_rows {
                let rows = if is_left { (&row, other_row) } else { (other_row, &row) };
                let row = self.make_row(&key, Some(rows.0), Some(rows.1), slicers, None);
                base.on_row(row)?;
            }
        }
        // put it in the store
        match this.1.entry(key) {
            Entry::Occupied(mut entry) => { entry.get_mut().push(row); },
            Entry::Vacant(entry) => { entry.insert(vec![row]); },
        }
        Ok(false)
    }
}

#[derive(Clone)]
struct Child {
    got_header: bool,
    left: bool,
    sender: Option<Sender<(bool, Message)>>,
}

impl Processor for Child {
    fn on_header(&mut self, _base: &mut Base, header: Vec<BString>) -> Result<()> {
        self.got_header = true;
        Break::when(self.sender.as_ref().unwrap().send((self.left, Message::Header(header))).is_err())
    }

    fn on_row(&mut self, base: &mut Base, row: Vec<BString>) -> Result<()> {
        if !self.got_header {
            self.on_header(base, vec![])?;
        }
        Break::when(self.sender.as_ref().unwrap().send((self.left, Message::Row(row))).is_err())
    }
}
